api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,302 @@
+// 2. The service sends back a MutationResponse message that directs Envoy
+//    to either stop processing, continue without it, or send it the
+//    next chunk of the message body.
+// 3. If so requested, Envoy sends the server chunks of the message body,
+//    or the entire body at once. In either case, the server sends back
+//    a MutationResponse after each message it receives.
+// 4. If so requested, Envoy sends the server the HTTP trailers,
+//    and the server sends back a MutationResponse.
+// 5. At this point, request processing is done, and we pick up again
+//    at step 1 when Envoy receives a response from the upstream server.
+// 6. At any point above, if the server closes the gRPC stream cleanly,
+//    then Envoy proceeds without consulting the server.
+// 7. At any point above, if the server closes the gRPC stream with an error,
+//    then Envoy returns a 500 error to the client, unless the filter
+//    was configured to ignore errors.
+//
+// In other words, the process is a request/response conversation, but
+// using a gRPC stream to make it easier for the client and server to
+// maintain state. The exception is the "SEND_BODY_ASYNC" command,
+// which requests that the server keep sending body chunks without waiting
+// for a response.
+
+service Mutation {
+  // This begins the bidirectional stream that Envoy will use to
+  // give the server control over what the filter does. The actual
+  // protocol is described by the MutationRequest and MutationResponse
+  // messages below.
+  rpc Mutate(stream MutationRequest) returns (stream MutationResponse) {
+  }
+}
+
+// This represents the different types of messages that Envoy can send
+// to a mutation server.
+// [#next-free-field: 8]
+message MutationRequest {
+  // If true, then Envoy requires that the server send one MutationResponse message
+  // back up the stream in response to this message -- otherwise, Envoy will
+  // never resume processing the filter chain. In practice, this will always
+  // be "true" unless the server or the user has configured the filter
+  // to run in "async mode."
+  bool response_required = 1;
+
+  // Information about the HTTP request, sent when the decodeHeaders callback
+  // is invoked on the Envoy filter.
+  Request request = 2;
+
+  // Information about the HTTP response, sent when the encodeHeaders callback
+  // is invoked on the Envoy filter.
+  Response response = 3;
+
+  // A chunk of the HTTP request body. This is only sent if the server
+  // requested it.
+  Body request_body = 4;
+
+  // A chunk of the HTTP response body, if the server requested it.
+  Body response_body = 5;
+
+  // The HTTP trailers for the request path. These are only sent if
+  // the server requested them.
+  config.core.v3.HeaderMap request_trailers = 6;
+
+  // The HTTP trailers for the response path, sent if the server
+  // requested them.
+  config.core.v3.HeaderMap response_trailers = 7;
```

Comment:
Should this be inside a oneof for clarity?

Reply:
Funny, because I just changed it from a oneof to a non-oneof.

In my mind each message will include only one of the above, but someone earlier suggested there might be a future version of the protocol in which the proxy can wait until it has all the bits and send them in only one message. For instance, maybe we have a configuration mode that sends the headers and the first body chunk in one message to optimize for a server that always needs to look at both before deciding what to do.